server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalCardinality.java

```diff
@@ -78,34 +78,30 @@ public long getValue() {
         return counts == null ? 0 : counts.cardinality(0);
     }
 
-    public HyperLogLogPlusPlus getCounts() {
+    public AbstractHyperLogLogPlusPlus getCounts() {
         return counts;
     }
 
     @Override
     public InternalAggregation reduce(List<InternalAggregation> aggregations, ReduceContext reduceContext) {
-        InternalCardinality reduced = null;
+        HyperLogLogPlusPlus reduced = null;
         for (InternalAggregation aggregation : aggregations) {
             final InternalCardinality cardinality = (InternalCardinality) aggregation;
             if (cardinality.counts != null) {
                 if (reduced == null) {
-                    reduced = new InternalCardinality(name, new HyperLogLogPlusPlus(cardinality.counts.precision(),
-                            BigArrays.NON_RECYCLING_INSTANCE, 1), getMetadata());
+                    reduced = new HyperLogLogPlusPlus(cardinality.counts.precision(),
+                        BigArrays.NON_RECYCLING_INSTANCE, 1);
```

Comment:
This means that as soon as you reduce you allocate the whole array, even if the cardinality is still small.

Reply:
Yes, the reduce phase is in the expensive object because the sparse does not know how to merge. Note it does not check for duplicates so it is not design for reduce, next iteration.